File: solutions/030-substring-with-concatenation-of-all-words/solution.py
from collections import Counter
from typing import List


class Solution:
    def find_substring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        word_len = len(words[0])
        word_count = len(words)
        target = Counter(words)
        result = []

        for offset in range(word_len):
            left = offset
            seen = Counter()
            used = 0

            for right in range(offset, len(s) - word_len + 1, word_len):
                word = s[right : right + word_len]
                seen[word] += 1
                used += 1

                while seen[word] > target.get(word, 0):
                    left_word = s[left : left + word_len]
                    seen[left_word] -= 1
                    if seen[left_word] == 0:
                        del seen[left_word]
                    left += word_len
                    used -= 1

                if used == word_count:
                    result.append(left)
                    left_word = s[left : left + word_len]
                    seen[left_word] -= 1
                    if seen[left_word] == 0:
                        del seen[left_word]
                    left += word_len
                    used -= 1

        return result


if __name__ == "__main__":
    tests = [
        ("barfoothefoobarman", ["foo", "bar"], [0, 9]),
        ("wordgoodgoodgoodbestword", ["word", "good", "best", "word"], []),
        ("barfoofoobarthefoobarman", ["bar", "foo", "the"], [6, 9, 12]),
    ]
    sol = Solution()
    for s, words, expected in tests:
        result = sorted(sol.find_substring(s, words))
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | s={s!r}, words={words} => {result} (expected {expected})")

File: solutions/030-substring-with-concatenation-of-all-words/solution.go
package main

import "fmt"

func findSubstring(s string, words []string) []int {
	if len(s) == 0 || len(words) == 0 {
		return nil
	}

	wordLen := len(words[0])
	wordCount := len(words)
	target := make(map[string]int, wordCount)
	for _, w := range words {
		target[w]++
	}

	result := []int{}

	for offset := 0; offset < wordLen; offset++ {
		left := offset
		used := 0
		seen := make(map[string]int)

		for right := offset; right <= len(s)-wordLen; right += wordLen {
			word := s[right : right+wordLen]
			seen[word]++
			used++

			for seen[word] > target[word] {
				leftWord := s[left : left+wordLen]
				seen[leftWord]--
				if seen[leftWord] == 0 {
					delete(seen, leftWord)
				}
				left += wordLen
				used--
			}

			if used == wordCount {
				result = append(result, left)
				leftWord := s[left : left+wordLen]
				seen[leftWord]--
				if seen[leftWord] == 0 {
					delete(seen, leftWord)
				}
				left += wordLen
				used--
			}
		}
	}

	return result
}

func main() {
	tests := []struct {
		s        string
		words    []string
		expected []int
	}{
		{"barfoothefoobarman", []string{"foo", "bar"}, []int{0, 9}},
		{"wordgoodgoodgoodbestword", []string{"word", "good", "best", "word"}, []int{}},
		{"barfoofoobarthefoobarman", []string{"bar", "foo", "the"}, []int{6, 9, 12}},
	}

	for _, tc := range tests {
		result := findSubstring(tc.s, tc.words)
		if !sameSlice(result, tc.expected) {
			fmt.Printf("FAIL | s=%q, words=%v => %v (expected %v)\n", tc.s, tc.words, result, tc.expected)
			continue
		}
		fmt.Printf("PASS | s=%q, words=%v => %v (expected %v)\n", tc.s, tc.words, result, tc.expected)
	}
}

func sameSlice(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	count := make(map[int]int)
	for _, v := range a {
		count[v]++
	}
	for _, v := range b {
		count[v]--
		if count[v] < 0 {
			return false
		}
	}
	return true
}
